readme.md
# Sensors

## Description

The purpose of the project it to demonstrate the implementation of a sensor API.

It is deployed to the following url:

http://sensors.kevin.colyar.net/v1/

## API Routes

### POST /temp

#### Request

The `POST` request at `/temp` should accepts a JSON blob in the following format:

- `{"data": __data_string__}`
- where `__data_string__` is format:
    - `__device_id__:__epoch_ms__:'Temperature':__temperature__`
      - where `__device_id__` is the device ID (int32)
      - `__epoch_ms__` is the timestamp in EpochMS (int64)
      - `__temperature__` is the temperature (float64)
      - and `'Temperature'` is the exact string

- Example `{"data": "365951380:1640995229697:'Temperature':58.48256793121914"}`

#### Response

- If for any reason the data string is not formatted correctly, return `{'error': 'bad request'}` with a `400` status code
- If the temperature is at or over 90
  - return `{'overtemp': True, 'device_id': __device_id__, 'formatted_time': __formatted_time__}`,
    - where `__device_id__` is the device ID (int32)
    - and `__formatted_time__` is the timestamp formatted to `%Y/%m/%d %H:%M:%S`
  - otherwise return `{'overtemp': False}`

### GET /errors

The `GET` request at `/errors` should return all data strings which have not been in the correct format.

### DELETE /errors

The `DELETE` request at `/errors` should clear the errors buffer.

### Live API Documentation

API Documentation can be viewed at the following url:

    http://sensors.kevin.colyar.net/v1/docs

## Releases/Roadmap

See `releases.md` for project progress.

## Development

The local API is deployed to the following url, per `docker-compose.yml`:

    open http://localhost:3001/v1/

## Deployment

Deploy to Digital Ocean kubernetes cluster via:

    ./bin/deploy.sh
    
Deployment settings can be found in `./k8/`

## Testing

### Test Suite

    docker compose run api pytest
    
### Test Coverage

    docker compose exec api coverage run -m pytest
    docker compose exec api coverage report -m
    docker compose exec api coverage html

    open htmlcov/index.html
    
### Command Line Testing

Select endpoint:

    export URL="http://localhost:3001"
    export URL="http://sensors.kevin.colyar.net"

Errors:

    curl $URL/v1/errors
    curl -X "DELETE" $URL/v1/errors

Temperature:

    curl -X POST -H "Content-Type: application/json" \
    -d '{"data": "365951380:1640995229697:'Temperature':58.48256793121914"}' \
    $URL/v1/temp

    curl -X POST -H "Content-Type: application/json" \
    -d '{"data": "365951380:1640995229697:'Temperature':90.0"}' \
    $URL/v1/temp

    curl -X POST -H "Content-Type: application/json" \
    -d '{"data": ":1640995229697:'Temperature':90.0"}' \
    $URL/v1/temp
    
### Smoke Test

A simple smoke test script of the production API is provided via:

    ./bin/smoke.sh

## Questions

Q: Why `int32` for `device_id`, not `uint32`? 

A: Using `int32` per spec, pending future changes.

Q: What unit is temperature in?

A: Assuming fahrenheit based on example data range (e.g. 58.5 & over 90 threshold).

Q: What timezone should `formatted_time` be returned in?

A: Assuming UTC since offset not specified in given format.


releases.md
# Releases

## v0.1.0 - Setup
- [X] Git
- [X] Python
- [X] Docker
- [X] Documentation
- [X] Testing
- [X] Logging

## v0.2.0 - Routes
- [X] Routes - POST /temp
- [X] Routes - GET /errors
- [X] Routes - DELETE /errors

## v0.3.x - Deployment
- [X] Deployment - k8
- [X] Deployment - digitalocean
- [X] Deployment - update pods after image push
  + https://stackoverflow.com/questions/33112789/how-do-i-force-kubernetes-to-re-pull-an-image
  + kubectl rollout restart deployments sensors-deploy
- [X] Deployment - dns - sensors.kevin.colyar.net

## v0.4.x
- [X] POST temp
- [X] Parse data
- [X] Response - {'overtemp': False}
- [X] Response - Error handling - Overtemp
- [X] Response - Error handling
- [X] Response - Format timestamp
- [X] Validation - device id - int32
- [X] Validation - epoch_ms - int64
- [X] Validation - temperature - float64

## v0.5.x
- [X] Testing - Setup
- [X] Testing - Unit tests
- [X] Testing - REST client
- [X] Testing - Bad type conversions
- [X] Testing - payload larger than VARCHAR 255

## v0.6.x
- [X] Persistence - schema
- [X] Persistence - stored procedures
- [X] Persistence - measurements
    + created_at
    + measurement/type
    + device_id 
    + value 
- [X] Persistence - errors
    + timestamp
    + route
    + method
    + payload
    + sort by timestamp

## v0.7.x
- [X] Documentation
  + https://www.linode.com/docs/guides/documenting-a-fastapi-app-with-openapi/
- [X] Documentation - API
  + https://fastapi.tiangolo.com/deployment/docker/#interactive-api-docs

## v0.8.x
- [X] API versioning

## v1.0.0
- [X] Push to github
- [X] Make github project public
- [X] Archive project
- [X] Email project

## v1.1.x
- [X] Linting
- [X] Coverage

## v1.2.x
- [ ] Persistence - test/staging db
- [ ] Persistence - indices
- [ ] Persistence - measurements - upsert
- [ ] Persistence - errors - save stack trace
- [ ] Persistence - handle db re/connection issues

## v1.3.x
- [ ] Lets Encrypt
- [ ] Kafka
- [ ] Authentication - JWT


main.py
#!/usr/bin/env python3

import logging
import dotenv
from fastapi import FastAPI, Query, Request
from fastapi.responses import JSONResponse
from fastapi_versioning import VersionedFastAPI, version
from pydantic import BaseModel
from typing import List

import exception_formatter
import sensors.commands as commands

# Initialization
# -----------------------------------------------------------------------------
dotenv.load_dotenv()

logger = logging.getLogger(__name__)
app = FastAPI()
db = commands.init_db()

# Types
# -----------------------------------------------------------------------------
class MeasurementRequest(BaseModel):
    data: str = None

class MeasurementResponse(BaseModel):
    overtemp: bool
    device_id: int
    formatted_time: str # TODO: provide format %Y/%m/%d %H:%M:%S

# Routes
# -----------------------------------------------------------------------------
@version(1)
@app.post("/temp")
@app.post("/measurement", response_model=MeasurementResponse)
async def measurement(
        measurement: MeasurementRequest,
        request: Request
):
    """
    Create a measurement record using the following format:

    - `{"data": __data_string__}`
    - where `__data_string__` is format:
    - `__device_id__:__epoch_ms__:__type__:__value__`, where:
      - `__device_id__` is the device ID (int32)
      - `__epoch_ms__` is the timestamp in EpochMS (int64)
      - `__type__` is the measurement type (string)
      - `__value__` is the value (float64)

    - __Note:__ Only measurement type of `'Temperature'` is supported at this time.

    - Example `{"data": "365951380:1640995229697:'Temperature':58.48256793121914"}`
    """
    try:
        return commands.save_measurement(db, measurement.data)

    except (ValueError, OverflowError, KeyError):
        logger.error(exception_formatter.format())

        commands.save_error(db, request.url._url, 'POST', measurement.data)

        return JSONResponse(
            status_code=400,
            content={'error': 'bad request'}
        )

@version(1)
@app.get("/errors", response_model=List[str])
async def get_errors():
    """
    Returns all data strings which have not been in the correct format.
    """
    return commands.get_errors(db)

@version(1)
@app.delete("/errors")
async def delete_errors():
    """
    Clears the error history.
    """
    commands.destroy_errors(db)
    return 'Errors cleared'

app = VersionedFastAPI(app, version_format='{major}', prefix_format='/v{major}')
